xcdat/dataset.py

```diff
@@ -0,0 +1,221 @@
+        Dataset with non-decoded CF/non-CF compliant time units.
+
+    Returns
+    -------
+    xr.Dataset
+        Dataset with decoded time units.
+
+    Notes
+    -----
+    .. [4] https://unidata.github.io/cftime/api.html#cftime.num2date
+
+    Examples
+    --------
+
+    Decode non-CF compliant time units in a Dataset:
+
+    >>> from xcdat.dataset import decode_time_units
+    >>> ds = xr.open_dataset("file_path", decode_times=False)
+    >>> ds.time
+    <xarray.DataArray 'time' (time: 3)>
+    array([0, 1, 2])
+    Coordinates:
+    * time     (time) int64 0 1 2
+    Attributes:
+        units:          years since 2000-01-01
+        bounds:         time_bnds
+        axis:           T
+        long_name:      time
+        standard_name:  time
+    >>> ds = decode_time_units(ds)
+    >>> ds.time
+    <xarray.DataArray 'time' (time: 3)>
+    array(['2000-01-01T00:00:00.000000000', '2001-01-01T00:00:00.000000000',
+        '2002-01-01T00:00:00.000000000'], dtype='datetime64[ns]')
+    Coordinates:
+    * time     (time) datetime64[ns] 2000-01-01 2001-01-01 2002-01-01
+    Attributes:
+        units:          years since 2000-01-01
+        bounds:         time_bnds
+        axis:           T
+        long_name:      time
+        standard_name:  time
+
+    View time coordinate encoding information:
+
+    >>> ds.time.encoding
+    {'source': None, 'dtype': dtype('int64'), 'original_shape': (3,), 'units':
+    'years since 2000-01-01', 'calendar': 'proleptic_gregorian'}
+    """
+    time = dataset["time"]
+    units_attr = time.attrs.get("units")
+
+    if units_attr is None:
+        raise KeyError(
+            "No 'units' attribute found for time coordinate. Make sure to open "
+            "the dataset with `decode_times=False`."
+        )
+
+    units, reference_date = units_attr.split(" since ")
+    non_cf_units_to_freq = {"months": "MS", "years": "YS"}
+
+    cf_compliant = units not in non_cf_units_to_freq.keys()
+    if cf_compliant:
+        dataset = xr.decode_cf(dataset, decode_times=True)
+        dataset.time.attrs["units"] = units_attr
+    else:
+        # NOTE: Calendar type for "months" or years" does not matter because the
+        # number of days in a month does not play a factor when generating date
+        # ranges for these units.
+        decoded_time = xr.DataArray(
+            data=pd.date_range(
+                start=reference_date,
+                periods=time.size,
+                freq=non_cf_units_to_freq[units],
+            ),
+            dims=["time"],
+            attrs=dataset["time"].attrs,
+        )
+        decoded_time.encoding = {
+            "source": dataset.encoding.get("source"),
+            "dtype": time.dtype,
+            "original_shape": decoded_time.shape,
+            "units": units_attr,
+            # pandas.date_range() returns "proleptic_gregorian" by default
+            "calendar": time.attrs.get("calendar", "proleptic_gregorian"),
+        }
```

Comment:
This might need a closer look. From my understanding, the number of days in a month should not matter if you are generating date ranges in months or years. Therefore, the calendar type has no influence on the operation.